src/ops.rs

```diff
@@ -619,16 +621,150 @@ impl AesEncryptable for Message {
 
         let key_schedule = AES::new(ke);
 
-        for block_index in (0..self.msg.len()).step_by(16) {
-            let temp = self.msg[block_index..block_index + 16].to_vec();
-            AES::decrypt_block(&mut self.msg, block_index, &key_schedule.round_key);
-            xor_blocks(&mut self.msg, &iv, block_index);
-            iv = temp;
-        }
-
+        let msg_copy = self.msg.clone();
+
+        self.msg.par_chunks_mut(16).enumerate().for_each(|(i, block)| {
+            let block_index = i * 16;
+            let xor_block = if block_index >= 16 {
+                &msg_copy[block_index - 16..block_index]
+            } else {
+                &iv // Use IV for the first block
+            };
+            // Decrypt the block in-place without using the output
+            AES::decrypt_block(block, 0, &key_schedule.round_key);
+            // XOR the decrypted block with the previous ciphertext block
+            xor_blocks(block, xor_block);
+        });
+    
         remove_pcks7_padding(&mut self.msg);
 
         let ver = &kmac_xof(ka, &self.msg, 512, "AES", 256);
         self.op_result = Some(self.digest.as_mut().unwrap() == ver);
     }
+
```

Comment:
Starting to wonder if there is a way to refactor aes in ops to keep common operations from being duplicated. As I write this I realize the same is probably true for the SHA3 encrypt/decrypt as well, maybe in the future we can refactor out the ke_ka stuff so we dont have to duplicate it, since ops.rs is getting so big now

Reply:
Thats a good point, it would definitely reduce the size and probably improve redability of ops.rs. I can look into that!